src/gitlab_bot.rs

```diff
@@ -0,0 +1,199 @@
+use crate::error::Error::GitlabApi;
+use crate::{error::*, Result};
+use curl::easy::Easy;
+use serde::Deserialize;
+use snafu::ResultExt;
+
+pub struct GitlabBot {
+	base_url: String,
+	ci_job_name: String,
+	private_token: String,
+}
+
+#[derive(PartialEq)]
+pub enum JobStatus {
+	Started,
+	AlreadyRunning,
+	Finished,
+	Unknown,
+}
+
+pub struct Job {
+	pub status: JobStatus,
+	pub status_raw: String,
+	pub url: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct GitlabJob {
+	id: i64,
+	name: String,
+	status: String,
+	web_url: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct Pipeline {
+	id: i64,
+}
+
+impl GitlabBot {
+	pub fn new_with_token(
+		hostname: String,
+		project: String,
+		ci_job_name: String,
+		private_token: String,
+	) -> Result<Self> {
+		let base_url = format!(
+			"https://{}/api/v4/projects/{}",
+			hostname,
+			project.replace("/", "%2F"),
+		);
+
+		// This request is just for checking that Gitlab is available and the token is valid.
+		get(&base_url, &private_token)?;
+
+		Ok(Self {
+			base_url: base_url.to_owned(),
+			ci_job_name: ci_job_name.to_owned(),
+			private_token: private_token.to_owned(),
+		})
+	}
+
+	pub fn build_artifact(&self, commit_sha: &str) -> Result<Job> {
+		let job = self.fetch_job(commit_sha)?;
+
+		// JobStatus is used by the caller to decide what message to post on Github/Matrix.
+		let status = match job.status.to_lowercase().trim() {
+			"manual" => JobStatus::Started,
+			"running" => JobStatus::AlreadyRunning,
+			"success" => JobStatus::Finished,
+			"failed" => JobStatus::Finished,
+			"canceled" => JobStatus::Finished,
+			_ => JobStatus::Unknown,
+		};
+
+		if status == JobStatus::Started {
+			let play_job_url =
+				format!("{}/jobs/{}/play", self.base_url, job.id);
+			let response = post(&play_job_url, &self.private_token)?;
+
+			if response.status > 299 {
+				return Err(Error::StartingGitlabJobFailed {
+					status: response.status,
+					url: job.web_url,
+					body: response.body,
+				});
+			}
+		}
+
+		Ok(Job {
+			status,
+			status_raw: job.status,
+			url: job.web_url.clone(),
+		})
+	}
+
+	fn fetch_job(&self, commit_sha: &str) -> Result<GitlabJob> {
+		let pipelines_url = format!(
+			"{}/pipelines?sha={}&order_by=updated_at&per_page=1",
+			&self.base_url, commit_sha,
+		);
+
+		let response = get(&pipelines_url, &self.private_token)?;
+		let pipelines: Vec<Pipeline> =
+			serde_json::from_str(&response.body).context(Json)?;
+
+		if pipelines.len() < 1 {
```

Comment:
```suggestion
		if pipelines.is_empty() {
```

Reply:
> > Mostly because I wanted to change as little of the existing code as possible. But now that I got the new code working, refactoring this a bit shouldn't be too hard. I'll give it a try.
> 
> @haikoschol feel free to ignore my comments for now. I do agree that a pull request should do at most one thing, thus involved refactorings should happen in a follow up pull request.

Right, I just decided not to change `handle_comment` in this PR since the benchmarking command will need to pass arguments. I'd rather do the concrete implementation for that together with the benchmarking command then write speculative code as part of a refactoring.